testing/example_scripts/issue_519.py

```diff
@@ -22,13 +22,13 @@ def checked_order():
     assert order == [
         ("issue_519.py", "fix1", "arg1v1"),
         ("test_one[arg1v1-arg2v1]", "fix2", "arg2v1"),
-        ("test_two[arg1v1-arg2v1]", "fix2", "arg2v1"),
```

Comment:
my understanding is, that this order ought not to change

Reply:
Looking at fixtureargkeys of these 8 tests, due to basing them on indices, currently we have:
```
<Function test_one[arg1v1-arg2v1]> {('arg1', 0, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v1-arg2v2]> {('arg1', 1, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v2-arg2v1]> {('arg1', 2, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v2-arg2v2]> {('arg1', 3, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v1-arg2v1]> {('arg1', 0, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v1-arg2v2]> {('arg1', 1, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v2-arg2v1]> {('arg1', 2, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v2-arg2v2]> {('arg1', 3, PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
```
in which `test_one[arg1v1-arg2v1]>` and `test_one[arg1v1-arg2v2]>` unexpectedly don't have a common dependency while they do depend on `arg1="arg1v1"`. This is also the case for `test_one[arg1v2-arg2v1]>` and `test_one[arg1v2-arg2v2]>`, etc.

By basing fixtureargkeys on values, we would have:
```
<Function test_one[arg1v1-arg2v1]> {('arg1', "arg1v1", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v1-arg2v2]> {('arg1', "arg1v1", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v2-arg2v1]> {('arg1', "arg1v2", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_one[arg1v2-arg2v2]> {('arg1', "arg1v2", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v1-arg2v1]> {('arg1', "arg1v1", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v1-arg2v2]> {('arg1', "arg1v1", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v2-arg2v1]> {('arg1', "arg1v2", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
<Function test_two[arg1v2-arg2v2]> {('arg1', "arg1v2", PosixPath('/tmp/pytest-of-sadra/pytest-0/test_5190/issue_519.py')): None}
```
in which `<Function test_one[arg1v1-arg2v1]>` correctly has common dependency with both `<Function test_one[arg1v1-arg2v2]>` and `<Function test_two[arg1v1-arg2v1]>`. This changes the reordering output thus affecting `test_issue519`. Am I correct?

Reply:
scope ordering gets broken up wrong

the parameter sets order is more important that the test name, so all test of a particular parameterization must come before the next test
so the order change here is incorrect and to be considered a bug

Reply:
`arg2=arg2v1` is not considered into reordering as its scope is `function`.

Reply:
I recall that it was added to allow for caching effects on repeated usages of a parameter, the idea being repeated Setup of the same key is commonly cheaper than going through different keys

Reply:
But we do not cache function-scoped parameters. Besides, the number of setups for the module-scoped fixture (arg1) has not changed.
I note that this PR hasn't any change in reordering algo. Just I've made fixture arg keys to first rely on param values if possible, as does `FixtureDef::cache_key`. This fixes problems like the one below
```
import pytest

@pytest.fixture(scope='module')
def fixture1(request):
    pass

@pytest.mark.parametrize("fixture1", [1, 0], indirect=True)
def test_0(fixture1):
    pass

@pytest.mark.parametrize("fixture1", [2, 1], indirect=True)
def test_1(fixture1):
    pass
```
in which we get inefficient ordering right now:
```
<Function test_0[1]>
<Function test_1[2]>
<Function test_0[0]>
<Function test_1[1]>
```

Reply:
The point in this case is not caching the value but the presumption of caching effects in the system when using the same Parameter for recreation

Reply:
I'm not sure if I understand, could you please explain a little bit or provide an example?

By the way, regardless of how should be test_issue519 output, do you agree that there's a discrepancy between what fixture argkey bases itself on (param index) and what fixturedef caches bases itself on (param value), resulting in inefficient orderings?

Reply:
But in that case and also in general, isn't better that user declare the parameter as a **high-scoped** one so that all tests reading a specific file, be close to each other?

Reply:
If repeated setup is needed that's neither intuitive nor what people do

Reply:
Then, what's the reason that user could set scope for direct parametrizations?

To set another example that basing arg keys on param index and `add_funcarg_pseudo_fixturedef`'s implementation produce unwanted results, by swapping `arg2` values for `test_two` we would have a whole different ordering

```python
import pytest

@pytest.mark.parametrize("arg2", ["arg2v1", "arg2v2"])
@pytest.mark.parametrize("arg1", ["arg1v1", "arg1v2"], scope='module')
def test_one(arg1, arg2):
    pass

@pytest.mark.parametrize("arg2", ["arg2v2", "arg2v1"])
@pytest.mark.parametrize("arg1", ["arg1v1", "arg1v2"], scope='module')
def test_two(arg1, arg2):
    pass
```
```
<Function test_one[arg1v1-arg2v1]>
  <Function test_two[arg1v1-arg2v2]>
  <Function test_one[arg1v1-arg2v2]>
  <Function test_two[arg1v1-arg2v1]>
  <Function test_one[arg1v2-arg2v1]>
  <Function test_two[arg1v2-arg2v2]>
  <Function test_one[arg1v2-arg2v2]>
  <Function test_two[arg1v2-arg2v1]>
```

Reply:
IMO the new order is more intuitive. I feel that having the same test definition (i.e. the base un-parametrized test) should have a stronger ordering affect than sharing a function-scoped parameterset. 

I understand @RonnyPfannschmidt comments re. the filesystem caching example, but I also agree with @sadra-barikbin that for such cases the user should set a higher scope.

Reply:
sometimes the user is not ableto set a higher scope (when using fixtures that have to recreate for each test)

the alternative would be to introduce a new fixture that is indeed of higher scope and passed in as the "parameter" for another fixture (which we currently cant do nicely)

if we can make a higher scope parameterize of a function scoped fixture work my concern would be obsoleted

let me reassure you that the new way proposed here is indeed more beautiful, more clean and more maintainable, but i'm under the impression it also breaks some "dirty wins"

as of now its unclear how to spell those "dirty wins cleanly, 

pytest-lazyfixtures has a potential way out by allowing fixtures to be spelled there and when integrating it in core parametrize support could be added



then we would be able to spell something like 

```python

inputs = pytest.fixture.from_constants([1,2,3,4], scope="module")

@pytest.fixture
def something(request):
   return request.param

@pytest.mark.parametrize('something', inputs, indirect=True)
def mytest(something):
   pass

```